[INTEL MKL] Enable MatMul in eager mode

gzmkl opened the pull request:
Enable MKL MatMul in eager mode.

Pending on the following PRs (that is, this PR should be merged after them):
https://github.com/tensorflow/tensorflow/pull/30401   (MKL conv fwd in eager mode)
https://github.com/tensorflow/tensorflow/pull/30402   (MKL conv bwd in eager mode)

d02k01 committed:
Copy tf_attrtype.h

d02k01 committed:
Copy tf_datatype.h, tf_status.h and tf_tensor.h

feihugis committed:
Refactor the test style for Datasets to reduce the repeat code

feihugis committed:
Add the test for IsStateful() API

feihugis committed:
Address the comments

gzmkl committed:
enable matmul in eager mode

agramesh1 commented:
Pinging @penpornk for review.

penpornk reviewed (commented):
This PR overlaps a lot with https://github.com/tensorflow/tensorflow/pull/30401. Can you make this PR depend on https://github.com/tensorflow/tensorflow/pull/30401? Otherwise, we should just wait until https://github.com/tensorflow/tensorflow/pull/30401 is merged.

penpornk commented:
(I meant, this PR seems to have just a few lines of changes when based on https://github.com/tensorflow/tensorflow/pull/30401.)

feihugis committed:
Modulize the TestCase

feihugis committed:
Add the utility function CreateTensors()

gzmkl commented:
@penpornk.  Yes, it depends on PR #30401. It is better waiting #30401 is merged. Thanks!

feihugis committed:
Add the missing virtual destructor

nhasabni committed:
Enable context-based graph rewrite

This PR enables context-based graph rewrite to make better decisions about
rewriting an op to MKL op. This PR also cleans up rewrite logic used for
element-wise ops and unifies it with context-based graph rewrite. Unit tests
are also added.

As a side note, a few changes are suggested by Clang format check, and are
not related to this PR.

bhavani-subramanian committed:
Reverting Add + Conv fusion change made during MKL-DNN v1.x integration.

gzmkl commented:
Headsup: after PR https://github.com/tensorflow/tensorflow/pull/30402 being merged, 
there will be some conflicts which I will resolve.

bhavani-subramanian committed:
Changed CHECK to DCHECK.

bhavani-subramanian committed:
Addressed review comments.

rxsang committed:
Automated rollback of commit 41884bf3d3c29f1520ba84a5233e30cb863f736e

PiperOrigin-RevId: 261406472

kkimdev committed:
Use ObjectIdentityDictionary() in session.py for Tensors

PiperOrigin-RevId: 261408384

dubey committed:
Configure `NcclGather` in collective param resolution.

Also add python tests that cover NCCL implementations of broadcast and
all-gather.

PiperOrigin-RevId: 261408514

edloper committed:
When constructing a RaggedTensor with a numpy row-partitioning tensor, don't convert int32->int64.

PiperOrigin-RevId: 261408643

lamberta committed:
Fix quant notebooks for tensorflow.org

PiperOrigin-RevId: 261408660

sun51 committed:
Only apply check for non-tensor case

PiperOrigin-RevId: 261408908

tensorflower-gardener committed:
Optimize 8bit Softmax op handler

PiperOrigin-RevId: 261410252

aaroey committed:
Update error message for TrtGraphConverter.

PiperOrigin-RevId: 261413236

Commit:
Delete the Windows CopyFileW definitions to fix the linker issues.

PiperOrigin-RevId: 261414544

tensorflower-gardener committed:
Merge pull request #31254 from Intel-tensorflow:nhasabni/context_based_graph_rewrite

PiperOrigin-RevId: 261404733

haozha111 committed:
Export model's minimum required runtime string into flatbuffer's metadata field. Also add a method in FlatBufferModel to parse the runtime string.

PiperOrigin-RevId: 261418624

pavithrasv committed:
Enable the Keras compile API `experimental_run_tf_function` flag by default.

This flag enables single training/eval/predict execution path. With this
1. All input types are converted to `Dataset`.
2. When distribution strategy is not specified this goes through the no-op distribution strategy path.
3. Execution is wrapped in tf.function unless `run_eagerly=True` is set in compile.

The single path execution code does not yet support all use cases. We fallback to the existing v1 execution paths if your model contains the following:
1. sample_weight_mode in compile
2. weighted_metrics in compile
3. ModelCheckpoint callback with `save_freq` set to integer
4. TensorBoard callback with `update_freq` set to integer
5. v1 optimizer
6. target tensors in compile

If you are experiencing any issues because of this change, please inform us (file an issue) about your use case and you can unblock yourself by setting `experimental_run_tf_function=False` in compile meanwhile.

We have seen couple of use cases where the model usage pattern is not as expected and would not work with this change.
1. output tensors of one layer is used in the constructor of another.
2. symbolic tensors outside the scope of the model are used in custom loss functions.
The flag can be disabled for these cases and ideally the usage pattern will need to be fixed.

PiperOrigin-RevId: 261418677

gzmkl committed:
Merge branch 'master' into eager_matmul_rewrite

gzmkl commented:
FYI - the expected merge conflicts have been resolved

kkimdev committed:
Use ObjectIdentity...() in def_function.py and func_graph.py for Tensors

PiperOrigin-RevId: 261419727

penpornk commented on tensorflow/core/common_runtime/eager/BUILD:
```diff
@@ -9,6 +9,12 @@ load(
     "mkl_deps",
 )
 
+load(
```
This has already been loaded in the lines above. Please remove.

penpornk commented on tensorflow/core/common_runtime/eager/BUILD:
```diff
@@ -272,6 +278,7 @@ cc_library(
             "//tensorflow/core/distributed_runtime/eager:remote_copy_node",
         ],
     }) + if_mkl([":mkl_eager_op_rewrite"]),
+
```
Please remove the extra blank line.

penpornk reviewed (changes requested):
Thank you! A few minor changes please.

aaroey committed:
Improve the calibration logic for TF 2.0 so it won't create the calibration
thread during inference (when input shapes don't match, calibration_mode_ is
true, and cache is not empty). Instead it'll now fallback to native function
execution.

PiperOrigin-RevId: 261421499

kkimdev committed:
Use ObjectIdentity...() in Keras backend.py for Tensors

PiperOrigin-RevId: 261421598

gharibian committed:
Added stand-in tstring with expected API.

See: https://github.com/tensorflow/community/pull/91
PiperOrigin-RevId: 261423971

tensorflower-gardener committed:
Merge pull request #31288 from Intel-tensorflow:bhavanis/revert-add-fusion

PiperOrigin-RevId: 261418872

saxenasaurabh committed:
Map TensorFlow StatelessWhile and While op to a common While op in MLIR

TensorFlow StatelessWhile and While op only differs in the is_stateful property and are identical otherwise. Introduced an additional attribute in the MLIR op definition to differentiate them and mapped to and from the common op while importing and export to MLIR, respectively.

PiperOrigin-RevId: 261428549

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 261428654

reedwm committed:
Ensure AutoCastVariable.dtype is not a string.

This is required for AutoCastVariable.__repr__ to work properly.

PiperOrigin-RevId: 261429675

Commit:
Fix the auditwheel repair flags for pip_new.sh to be manylinux2010 compliant.

PiperOrigin-RevId: 261430480

Commit:
Fix cond_v2 and while_v2 to be consistent with Eager/NumPy: nonscalar predicates are allowed, so long as they have a single element.

PiperOrigin-RevId: 261433568

saxenasaurabh committed:
Enable v2 control flow inside graph_mode in TF 2.0.
Current behavior is to use v2 control flow only inside tf.function.

PiperOrigin-RevId: 261435951

Commit:
Extend unit tests for small keras models using a variety of layers and parameters.

PiperOrigin-RevId: 261438511

jingpu committed:
Extend RankOp::fold to also fold when input has a known rank.

Currently, it does not fold if the rank is zero because of b/138865275.

PiperOrigin-RevId: 261439262

tensorflower-gardener committed:
Merge pull request #31059 from ocjosen:patch-1

PiperOrigin-RevId: 261444666

tensorflower-gardener committed:
Disable test in v2 builds.

PiperOrigin-RevId: 261449859

gunan committed:
Automated rollback of commit a966b1743ed0df7b22d35de27cb2203767b110d6. Revert #21823.

PiperOrigin-RevId: 261453723

karimnosseir committed:
Add ResizeNearestNeighbor op to MLIR Converter

PiperOrigin-RevId: 261456650

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2019-08-03

PiperOrigin-RevId: 261462124

tensorflower-gardener committed:
Update GraphDef version to 116.

PiperOrigin-RevId: 261462127

sun51 committed:
Only apply check for non-tensor case

PiperOrigin-RevId: 261481490

jsimsa committed:
[tf.data] Reducing verbosity of DirectedInterleaveDataset warning.

PiperOrigin-RevId: 261483964

tensorflower-gardener committed:
Merge pull request #31105 from feihugis:Tests_IsStateful

PiperOrigin-RevId: 261483873

agramesh1 committed:
Merge branch 'eager_matmul_rewrite' of https://github.com/Intel-tensorflow/tensorflow into eager_matmul_rewrite

agramesh1 committed:
Fixed merge issues

agramesh1 commented:
Please ignore, Git merge messed things up. Created a new PR here https://github.com/tensorflow/tensorflow/pull/31311  I will try to close this PR.

penpornk commented:
I'll close the PR for you.
